File: src/main/kotlin/leetcode/FoodRatings.kt
package leetcode

import java.util.*
import java.util.function.Function


data class Food(val foodRating:Int,val foodName:String) : Comparable<Food>{
    override fun compareTo(other: Food): Int {
        if (foodRating==other.foodRating){
            return foodName.compareTo(other.foodName)
        }
        return -1 * foodRating.compareTo(other.foodRating)
    }

}
class FoodRatings(foods: Array<String>, cuisines: Array<String>, ratings: IntArray) {
    val foodRatingMap= mutableMapOf<String,Int>()
    val foodCuisineMap= mutableMapOf<String,String>()
    val cuisineFoodMap= mutableMapOf<String,PriorityQueue<Food>>()

    init {
        for (i in foods.indices) {
            // Store 'rating' and 'cuisine' of the current 'food' in 'foodRatingMap' and 'foodCuisineMap' maps.
            foodRatingMap[foods[i]] = ratings[i]
            foodCuisineMap.put(foods[i], cuisines[i])
            // Insert the '(rating, name)' element into the current cuisine's priority queue.
            cuisineFoodMap.computeIfAbsent(
                cuisines[i]
            ) { _ -> PriorityQueue<Food>() }.add(
                Food(
                    ratings[i],
                    foods[i]
                )
            )
        }
    }
    fun changeRating(food: String, newRating: Int) {
        foodRatingMap[food] = newRating
        val cuisineName= foodCuisineMap[food]
        cuisineFoodMap[cuisineName]?.add(Food(newRating,food))
    }

    fun highestRated(cuisine: String): String {
        var cuisineHighRate= cuisineFoodMap[cuisine]?.peek()

        while (foodRatingMap[cuisineHighRate?.foodName?:""] != cuisineHighRate?.foodRating){
            cuisineFoodMap.getValue(cuisine).poll()
            cuisineHighRate=cuisineFoodMap[cuisine]?.peek()
        }
        return cuisineHighRate?.foodName?:""
    }

}

File: src/main/python/leetcode/FoodRatings.py
class FoodRatings:
    def __init__(self, foods: List[str], cuisines: List[str], ratings: List[int]):
        self.cuisines = defaultdict(list)
        self.foodRatings = {}
        self.foodCuisines = {}
        for food, cuisine, rating in zip(foods, cuisines, ratings):
            self.foodRatings[food] = rating
            self.foodCuisines[food] = cuisine
            heapq.heappush(self.cuisines[cuisine], (-rating, food))

    def changeRating(self, food: str, newRating: int) -> None:
        self.foodRatings[food] = newRating
        cuisine = self.foodCuisines[food]
        heapq.heappush(self.cuisines[cuisine], (-newRating, food))

    def highestRated(self, cuisine: str) -> str:
        heap = self.cuisines[cuisine]
        while self.foodRatings[heap[0][1]] != -heap[0][0]:
            heapq.heappop(heap)
        return heap[0][1]
